### PR TITLE
feat(packages): use react 17

### LINKED ISSUE #5077: Support for React 17 in PatternFly 3?
I tried to update to React 17.0.1 and it seems that at least menu items in the masthead breaks (on click and drop down menus dont work). I believe this is due to the change of event logic with React 17, but I have not been able to locate the specific problem - I guess it also could be Bootstrap. I'm using latest PatternFly 3 release.

Any plans on supporting React 17 in PatternFly 3?

### LINKED ISSUE #5108: React peer dependency version restriction
When using npm@7 and react 17.x PF throws invalid peer dependency error (npm@7 will try to install all peer dependencies).

Is it possible to change the version restriction of React peer dependency from `^16.8.x` to `>=16.8.x`? As far as I can tell PF4 works with react 17.

You can get around this error by using `--force` or use legacy peer deps resolution, but that can lead to issues with other packages.

### PR DESCRIPTION
**Pull request by redallen**
**What**:

Updates peer dependencies for React to `^16.8.0 || ^17.0.0`. Upgrades React in the lerna root, react-docs, and demo-app-ts to React 17.

Closes #5077.
Closes #5108.

**Additional issues**: Using unofficial enzyme adapter as a workaround until React 17 support lands upstream. Issue: https://github.com/enzymejs/enzyme/issues/2429.

### CODE REVIEW

**Comment by evwilkin** [comment]
@redallen merge conflicts

File: packages/react-core/src/components/Dropdown/Toggle.tsx [hunk 1]
```diff
@@ -104,6 +104,7 @@ export class Toggle extends React.Component<ToggleProps> {
       return;
     }
     if (!this.props.bubbleEvent) {
+      this.onEscPress((event as unknown) as KeyboardEvent);
```
Comment (jschuler) [inline packages/react-core/src/components/Dropdown/Toggle.tsx:107]: can you describe the issue you were seeing

Comment (redallen) [reply packages/react-core/src/components/Dropdown/Toggle.tsx:107]: The escape key was not closing our Dropdown component. On React 16 it was.

Comment (jschuler) [reply packages/react-core/src/components/Dropdown/Toggle.tsx:107]: have you tried the method described here?
https://reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues
i.e. adding `{capture: true}` to the `document.addEventListener('keydown', this.onEscPress);`

Comment (redallen) [reply packages/react-core/src/components/Dropdown/Toggle.tsx]: That works as well and I've updated the PR to use that instead.

**Comment by redallen** [comment]
We should recheck components with global event listeners per @jschuler 's reccomendation. These are the files that use `.addEventListener`:
- packages/react-charts/src/components/ChartArea/examples/ChartArea.md:
- packages/react-charts/src/components/ChartBullet/examples/ChartBullet.md:
- packages/react-charts/src/components/ChartLegend/examples/ChartLegend.md:
- packages/react-charts/src/components/ChartLine/examples/ChartLine.md:
- packages/react-charts/src/components/ChartScatter/examples/ChartScatter.md:
- packages/react-charts/src/components/ChartStack/examples/ChartStack.md:
- packages/react-charts/src/components/ChartThreshold/examples/ChartThreshold.md:
- packages/react-console/src/components/SerialConsole/XTerm.tsx:
- packages/react-console/src/components/VncConsole/VncConsole.tsx:
- packages/react-core/src/components/AboutModal/AboutModal.tsx:
- packages/react-core/src/components/AboutModal/__tests__/AboutModal.test.tsx:
- packages/react-core/src/components/ContextSelector/ContextSelectorToggle.tsx:
- packages/react-core/src/components/Drawer/DrawerPanelContent.tsx:
- packages/react-core/src/components/Dropdown/DropdownMenu.tsx:
- packages/react-core/src/components/Dropdown/Toggle.tsx:
- packages/react-core/src/components/Dropdown/__tests__/DropdownToggle.test.tsx:
- packages/react-core/src/components/DualListSelector/DualListSelector.tsx:
- packages/react-core/src/components/DualListSelector/DualListSelectorPane.tsx:
- packages/react-core/src/components/JumpLinks/JumpLinks.tsx:
- packages/react-core/src/components/Modal/Modal.tsx:
- packages/react-core/src/components/Modal/__tests__/Modal.test.tsx:
- packages/react-core/src/components/Nav/NavList.tsx:
- packages/react-core/src/components/OptionsMenu/OptionsMenuToggleWithText.tsx:
- packages/react-core/src/components/OverflowMenu/OverflowMenu.tsx:
- packages/react-core/src/components/Page/Page.tsx:
- packages/react-core/src/components/Popover/PopoverCloseButton.tsx:
- packages/react-core/src/components/Select/SelectToggle.tsx:
- packages/react-core/src/components/Select/__tests__/SelectToggle.test.tsx:
- packages/react-core/src/components/Tabs/Tabs.tsx:
- packages/react-core/src/components/TextInput/TextInput.tsx:
- packages/react-core/src/components/Toolbar/Toolbar.tsx:
- packages/react-core/src/components/Toolbar/examples/Toolbar.md:
- packages/react-core/src/components/Wizard/Wizard.tsx:
- packages/react-core/src/helpers/Popper/Popper.tsx:
- packages/react-core/src/helpers/Popper/thirdparty/popper-core/modifiers/eventListeners.ts:
- packages/react-integration/demo-app-ts/src/components/demos/ToolbarDemo/ToolbarDemo.tsx:
- packages/react-table/src/components/Table/RowWrapper.tsx:
- packages/react-topology/src/behavior/usePanZoom.tsx:
- packages/react-topology/src/components/popper/Popper.tsx:
- packages/react-topology/src/hooks/useEventListener.tsx:
- packages/react-topology/src/layouts/BaseLayout.ts:
- packages/react-topology/src/utils/useHover.ts:
- packages/react-virtualized-extension/src/components/Virtualized/examples/VirtualizedTable.md:
- packages/react-virtualized-extension/src/components/Virtualized/examples/WindowScroller.md:

**Comment by boaz0** [comment]
I am so happy we are having React v17 update in mind!
Good luck and well done :+1:

**Comment by redallen** [comment]
I've fixed the merge conflicts and checked that I can't easily break the keyboard navigation of each of the react-core components listed in my last comment. I'll leave the rest to reviewers like @jessiehuff .

**Review by jessiehuff — DISMISSED** [review]
LGTM! 🙂  I tested every item that I could on your list and didn't notice any issues. We don't currently have a lot of keyboard interaction on our charts so those seem unaffected from what I can tell. The other extensions that I could test look the same as what we currently have as well. There's a11y concerns with those, but none new to this PR and already have open issues.

File: package.json [hunk 2]
```diff
@@ -58,8 +58,8 @@
     "mutation-observer": "^1.0.3",
     "plop": "^2.0.0",
     "prettier": "^1.18.2",
-    "react": "^16.4.0",
-    "react-dom": "^16.4.0",
+    "react": "^17.0.0",
```
Comment (tlabaj) [inline package.json:61]: @redallen  will this force consumers to upgrade there version of react?

**Review by tlabaj — APPROVED** [review]

**Comment by redallen** [comment]
@tlabaj It won't force consumers to upgrade because the peer dependency is changing to:
```json
"react": "^16.8.0 || ^17.0.0",
"react-dom": "^16.8.0 || ^17.0.0"
 ```

**Review by jessiehuff — APPROVED** [review]

**Review by evwilkin — APPROVED** [review]
LGTM

**Pull request merged**